---
title: "FusionExpressionPlot Walkthrough"
author: "Stuart R. Jefferys"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{FusionExpressionPlot Walkthrough}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Set up to run
=============

1. Make working directory and work there

        mkdir /home/srj/Projects/Fusions/Pcpg/FusionExpression
        cd /home/srj/Projects/Fusions/Pcpg/FusionExpression
        
1. Getting list of quantification files:

        psql -U seqware -h swprod.bioinf.unc.edu seqware_meta_db --field-separator $'\t' --no-align --tuples-only --command "select v.sample, v.file_path from vw_files v, upload u where u.sample_id = v.sample_id and v.sample_type = 'PCPG' and v.algorithm = 'bt_exon_quantification' and (u.target='CGHUB' or u.target='CGHUB_BAM') and u.external_status = 'live' and (u.status != 'skip')" -o exonQuantFileList.txt
        
1. Verify no sample listed twice by comparing:

        cut -f 1 exonQuantFileList.txt | sort | uniq | wc -l
        wc -l exonQuantFileList.txt

1. Copy R code files

        cp /home/srj/Projects/Fusions/src/ExonExpressionPlots_0.1/*.R .
        
1. Get list of samples in cohort

        cut -f 1 exonQuantFileList.txt > sampleList.txt

1. Add heading to exonQuantFileList.txt

        vi exonQuantFileList.txt
            1G
        i
        samples	exon_expression_files

        :wq
        
1. Create list of genes to plot (genesList.txt) with one gene per line

1. Get Mapsplice cohort fusion list and save locally

Get gene model
==============

1. Link to gaf file

        ln -s /datastore/tier1data/nextgenseq/seqware-analysis/GAF/TCGA.hg19.June2011.gaf TCGA.hg19.June2011.gaf
        
1. In R with this as the working directory

    1. Set up names for input, output, and internal files
    
            gafFile <- "TCGA.hg19.June2011.gaf"
            gafGeneModelsFile <- "TCGA.hg19.June2011.gaf.geneModels.txt"
            geneModels.Rdata <- "geneModels.RData"
            
    1.Load code to process the gaf
    
            source( "gafGeneModels.R" )
            
    1. Extract gene models from the gaf (takes a few seconds)
    
            gafGeneModelStats <- extractGeneModels( gafFile, outFile=gafGeneModelsFile )
            print(gafGeneModelStats)
                $gaf
                [1] "TCGA.hg19.June2011.gaf"
                
                $gaf_real
                [1] "/datastore/tier1data/nextgenseq/seqware-analysis/GAF/TCGA.hg19.June2011.gaf"
                
                $gaf_md5
                            TCGA.hg19.June2011.gaf
                "b9e0c2b81736d82d62bb6ab8cc517644"
                
                $uniqueGene
                [1] TRUE
                
                $skipUnknownGene
                [1] TRUE
                
                $gaf_lines
                [1] 3483704
                
                $gaf_models
                [1] 20764
                
                $gaf_models_unique
                [1] 20502
                
                $gaf_extract
                [1] "TCGA.hg19.June2011.gaf.geneModels.txt"
                
                $gaf_extract_real
                [1] "/home/srj/Projects/Fusions/Pcpg/FusionExpression/TCGA.hg19.June2011.gaf.geneModels.txt"
                
                $gaf_extract_md5
                TCGA.hg19.June2011.gaf.geneModels.txt
                   "4e048d450089ee2309dd16c6ce19b898"
                   
      . Generate the data file of gene models (takes a few seconds)

            geneModels <- loadGeneModels( gafGeneModelsFile )
            save( geneModels, file=geneModels.Rdata )
            
Generate cohort expression data frame
=====================================

1. Setup parameters for running
    
        expressionColumn <- "rpkm"
        geneModels.Rdata <- "geneModels.RData"
        cohortSamplesFile <- "sampleList.txt"
        sampleExpressionFileMap <- "exonQuantFileList.txt"
        cohortExpression.RData <- "cohortExpression.RData"
        
1. Load code to process the expression files

        source("cohortNormalizedExpression.R")
        
1. Load the gene model data prepared above

        load( geneModels.Rdata )

1. Set up the data formats used for processing gene expression files

        cohortSamples <- getSampleCohort( cohortSamplesFile )
            Read 184 items
        sampleExpressionFiles <- getCohortFiles( sampleExpressionFileMap, cohortSamples )
        
1. Load data from all cohort expression files, takes about 5 seconds per file

    cohortExpression <- getCohortExonExpressionData( geneModels, sampleExpressionFiles, type=expressionColumn )
            [1] "1 of 184. Adding exon expression data for TCGA-P7-A5NY-01A-12R-A35K-07"
            [1] "2 of 184. Adding exon expression data for TCGA-QR-A6H4-01A-11R-A35K-07"
            ...
            [1] "183 of 184. Adding exon expression data for TCGA-WB-A81K-01A-11R-A35L-07"
            [1] "184 of 184. Adding exon expression data for TCGA-S7-A7X0-01A-12R-A35L-07"
    save( cohortExpression, file=cohortExpression.RData )
        
Normalize
=========

1. Set up for normalization

        cohortExpression.RData <- "cohortExpression.RData"
        normalizedCohortExpression.RData <- "normalizedCohortExpression.RData"

1. Do normalization (Takes a couple of seconds per sample)

        normalizedCohortExpression <- normExpressionData( cohortExpression )
        save( normalizedCohortExpression, file=normalizedCohortExpression.RData )

Get Fusions to plot
===================

    fusionFile <- "pcpg_185_2014-05-18.merged.fusion_gene_profiles.txt"
    cohortSamplesFile <- "sampleList.txt"
    genesOfInterestFile <- "genesList.txt"
    fusionData.RData <- "fusionData.RData"
    selectedFusions.RData <- "selectedFusions.RData"
    source("cohortFusions.R")
    source("cohortNormalizedExpression.R")
    fusionData <- getMapSpliceCohortFusionData( fusionFile );
    save( fusionData, file=fusionData.RData )
    cohortSamples <- getSampleCohort( cohortSamplesFile )
        Read 184 items
    genesOfInterest <- getSampleCohort( genesOfInterestFile )
        Read 3 items
    selectedFusions <- filterFusions( fusionData, sample=cohortSamples, gene1=genesOfInterest, gene2=genesOfInterest );
    save( selectedFusions, file=selectedFusions.RData )

Make plots
==========

    selectedFusions.RData <- "selectedFusions.RData"
    normalizedCohortExpression.RData <- "normalizedCohortExpression.RData"
    source("cohortFusions.R")
    source("fusionExpressionPlot.R")
    load( selectedFusions.RData )
    load( normalizedCohortExpression.RData )
    do.allPlots( selectedFusions, normalizedCohortExpression)

